enhancements/hypershift/support-for-image-registry-capability.md

```diff
@@ -0,0 +1,476 @@
+Registry, and for that a pod redeployment is necessary.
+
+We propose a way to disable the Image Registry that conforms to
+[Hypershift's design principles](hypershift-design-principles).
+
+Introducing a capabilities API to Hypershift is the natural way of supporting
+disabling optional components at install time. To constrain the scope of the
+inital implementation, the `ImageRegistry` capability is the only one that will
+be initially supported, although the capabilities API itself will allow room for
+growth.
+
+[hypershift-design-principles]: https://hypershift-docs.netlify.app/reference/goals-and-design-invariants/
+
+
+### User Stories
+
+See [Hypershift docs](hypershift-personas) for a definition of the personas used below.
+
+* As a cluster instance admin, I want to provision hosted control planes and
+clusters without the Image Registry, so that my hosted clusters do not contain
+its resources, such as workloads, storage accounts, pull-secrets, etc
+
+[hypershift-personas]: https://hypershift-docs.netlify.app/reference/concepts-and-personas/#personas
+
+### Goals
+
+* Cluster instance admins can create clusters with the Image Registry disabled
+* Cluster instance admins can reenable the Image Registry without recreating
+the cluster
+* Components in the hosted control plane are not impacted by the Image Registry
+absence.
+* Image Registry related assets such as workloads, storage account and
+pull-secrets are not provisioned in clusters without the Image Registry
+* OpenShift Engineers are able to extend the list of capabilities available
+in Hypershift, allowing other optional components to be disabled in the future.
+
+### Non-Goals
+
+* Supporting the full set of OpenShift cluster capabilities
+* Moving management of hosted control plane components to CVO
+([OTA-951](https://issues.redhat.com/browse/OTA-951))
+
+## Proposal
+
+`HostedCluster` grows an optional array field `disabledCapabilities` in its `spec`,
+indicating what components users want to disable at install time.
+
+Initially, the only supported value in this array is `ImageRegistry`.
+
+Editing this field at runtime is only allowed if the user is removing
+capabilities from the array. Adding capabilities to the array is not supported.
+In other words, disabling capabilities at runtime is not supported.
+
+The `control-plane-operator` is responsible for reconciling both, the Image
+Registry and Cluster Version Operator. Both reconcilers will need to change.
+
+The Image Registry reconciler needs to look for an entry matching the
+`ImageRegistry` capability in the `HostedCluster.spec.disabledCapabilities`
+field, and skip reconciliation when the `ImageRegistry` capability is found.
+
+The `CVO` reconciler needs to calculate the difference between
+`HostedCluster.spec.disabledCapabilities` and
+`ClusterVersionCapabilitySets[ClusterVersionCapabilitySetCurrent]`, and assign
+the resulting set to field `spec.capabilities.baselineCapabilitySet` in the
```

Comment:
I think I got this one backward. You apparently have to set the `baselineCapabilitySet` to None.

```
spec:
  clusterID: %s
  capabilities:
    baselineCapabilitySet: 'None'
    additionalEnabledCapabilities: [ALL but ImageRegistry]
```

Alternatively, and maybe cleaner, we can introduce a Hypershift capability set just for this usecase

Reply:
source, if anyone else wants to look into it 
https://github.com/openshift/hypershift/pull/5315/files

Reply:
Nice catch, you are right: https://docs.openshift.com/container-platform/4.17/installing/overview/cluster-capabilities.html - I'll update this accordingly.
As for introducing a Hypershift specific capability set, I think that will end up constraining what can be disabled in the future. If we go for the on the fly calculation (base capability set - image registry capability = final capability set) then we open up for extending the set of disabled capabilities in the future.
The one question that remains is whether `ClusterVersionCapabilitySets[ClusterVersionCapabilitySetCurrent]` is the correct base capability set.
We must also find a good way to constraint the set of disabled capabilities to only those supported by Hypershift. I know we can use CEL validation for this, but I'm uncertain whether we want to constrain this in more places in the code.